### ISSUE TITLE
Error with AVIF version 0.10.1 & darktable, and avifImageYUVToRGB library

### ISSUE THREAD

**Issue by TopRockPhoto** [AOMediaCodec/libavif#1085]
avifImageYUBToRGB causing errors in Darktable and GIMP with AVIF 0.10.1

I was running darktable 4.0.0 & GIMP 2.10.30 with AVIF 0.9.3 and all was good. (I had compiled my own darktable).

I upgraded AVIF to 0.10.1 and then darktable would crash when attempting to export to an AVIF file. Tried GIMP with the same issue.

I tried to re-compile darktable with the new libraries and got…
`[ 34%] Building C object bin/CMakeFiles/lib_darktable.dir/common/imageio_avif.c.o
/home/karim/Downloads/AV/Photography/darktable/darktable/src/common/imageio_avif.c: In function ‘dt_imageio_open_avif’:
/home/karim/Downloads/AV/Photography/darktable/darktable/src/common/imageio_avif.c:74:12: error: too few arguments to function ‘avifImageYUVToRGB’
   74 |   result = avifImageYUVToRGB(avif, &rgb);
      |            ^~~~~~~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
make[2]: *** [bin/CMakeFiles/lib_darktable.dir/build.make:1947: bin/CMakeFiles/lib_darktable.dir/common/imageio_avif.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:2593: bin/CMakeFiles/lib_darktable.dir/all] Error 2
make: *** [Makefile:171: all] Error 2
`

Downgraded to AVIF ver 0.9.3 and everything works.

Ubuntu 22.04.1
GNOME 42.4.0
Linux Kernel 5.15.0-47-generic
GCC 11 (x86_64-linux-gnu)
CPU: AMD Ryzen 7 3800X 8-Core Processor
RAM: 32 GiB
GPU: NVIDIA UNIX x86_64 Kernel Module  470.141.03  Thu Jun 30 18:45:31 UTC 2022
FLAGS: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es

**Comment by wantehchang** [comment]
Hi Karim,

The function prototype of avifImageYUVToRGB() changed on the tip of libavif's main branch. You need to add AVIF_YUV_TO_RGB_DEFAULT as the third argument.

To help us invesitigate the crash of Darktable and GIMP with libavif 0.10.1, could you provide a stack trace of the crash? Thanks.

**Comment by kmilos** [comment]
@wantehchang Please keep a backward compatible 2-argument prototype that wraps this new call w/ default value. Having to do ifdefs on libavif version in many dependent packages would not be that cool.

**Comment by wantehchang** [comment]
Miloš: Thank you for your comment. We will look into this. Please review the following to make sure I understand your suggestion exactly:

1. Rename the three-parameter functions so that their names are different from the original functions names `avifImageRGBToYUV` and `avifImageYUVToRGB`. For example, we can rename them `avifImageRGBToYUVWithFlags` and `avifImageYUVToRGBWithFlags`.

2. Add back the two-parameter functions with the original function names `avifImageRGBToYUV` and `avifImageYUVToRGB`. The two-parameter functions call the three-parameter functions with default values for the `flags` parameters.

In pseudo code, the suggestion change is:

```
// New conversion functions.
AVIF_API avifResult avifImageRGBToYUVWithFlags(avifImage * image, const avifRGBImage * rgb,
                                               avifRGBToYUVFlags flags);
AVIF_API avifResult avifImageYUVToRGBWithFlags(const avifImage * image, avifRGBImage * rgb,                                               
                                               avifYUVToRGBFlags flags);

// Backward-compatible wrapper functions
AVIF_API avifResult avifImageRGBToYUV(avifImage * image, const avifRGBImage * rgb)
{
    return avifImageRGBToYUVWithFlags(image, rgb, AVIF_RGB_TO_YUV_DEFAULT);
}

AVIF_API avifResult avifImageYUVToRGB(const avifImage * image, avifRGBImage * rgb)
{
    return avifImageYUVToRGBWithFlags(image, rgb, AVIF_YUV_TO_RGB_DEFAULT);
} 
```

Is this what you suggested?

Note that the backward-compatible wrapper functions can only be used if the client code does **not** use the `avifChromaUpsampling chromaUpsampling` member of the `avifRGBImage` struct. In the packages you maintain, is this the case?

**Comment by wantehchang** [comment]
Miloš: I forgot to mention that we are working towards a stable API for the v1.0.0 release. This is why we changed the function prototypes of `avifImageRGBToYUV` and `avifImageYUVToRGB`.

Specifically, the `avifRGBImage` struct is designed to be allocated by the clients of libavif. In a stable API we won't be able to modify the members of a struct allocated by client code. So we decided to move the `avifRGBImage` struct members that are likely to change out of the struct. This necessitated passing the equivalent information as new input parameters to the conversion functions. The new input parameters are defined as a bitwise OR of bit flags to allow new flags to be added without changing the function prototypes.

**Comment by kmilos** [comment]
Thanks for taking the time to explain the API changes Wan-Teh. Indeed, as it would seem they're far-reaching, the client code change would also seem imminent after 1.0 (darktable might be an exception currently not using `chromaUpsampling`) so doubling functions like this (I forgot overloading is not an option here) might not be that pretty either. I'll let you make the decision.

As the original bug report goes, I guess somebody jumped the gun with shipping the main branch...

**Comment by TopRockPhoto** [comment]
Yes, I did realise that I went for the latest, listed as 0.10.1, but may
have taken master instead. I have been away from my desktop since posting
this bug report. Will try again tonight, when I get home, with the proper
source code.

Sorry about that mix-up with master.

However, am I to understand, that when 1.0.0 is released, there will be a
change of which developers such as darktable & Gimp would have to be aware?
